pow/src/main/java/tech/pegasys/artemis/DepositRequestManager.java

```diff
@@ -0,0 +1,107 @@
+package tech.pegasys.artemis;
+
+import com.google.common.primitives.UnsignedLong;
+import io.reactivex.disposables.Disposable;
+import org.apache.tuweni.bytes.Bytes32;
+import org.web3j.protocol.Web3j;
+import org.web3j.protocol.core.DefaultBlockParameter;
+import org.web3j.protocol.core.DefaultBlockParameterName;
+import org.web3j.protocol.core.methods.response.EthBlock;
+import tech.pegasys.artemis.pow.contract.DepositContract;
+import tech.pegasys.artemis.pow.event.Deposit;
+import tech.pegasys.artemis.pow.event.DepositsFromBlockEvent;
+import tech.pegasys.artemis.util.async.SafeFuture;
+
+import java.math.BigInteger;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static tech.pegasys.artemis.util.config.Constants.ETH1_FOLLOW_DISTANCE;
+
+public class DepositRequestManager {
+
+  private final Web3j web3j;
+  private final Consumer<DepositsFromBlockEvent> depositsFromBlockEventConsumer;
+  private final Map<String, SafeFuture<EthBlock.Block>> cachedBlocks = new ConcurrentHashMap<>();
+
+  private BigInteger latestSuccessfullyQueriedBlockNumber = BigInteger.ZERO;
+  private Disposable newBlockSubscription;
+  private BiFunction<
+          DefaultBlockParameter,
+          DefaultBlockParameter,
+          List<DepositContract.DepositEventEventResponse>> depositEventsInRangeFunction;
+
+  public DepositRequestManager(Web3j web3j,
+                               Consumer<DepositsFromBlockEvent> depositsFromBlockEventConsumer,
+                               BiFunction<
+                                       DefaultBlockParameter,
+                                       DefaultBlockParameter,
+                                       List<DepositContract.DepositEventEventResponse>> depositEventsInRangeFunction) {
```

Comment:
I'd suggest passing in the actual `DepositContract` here rather than a `BiFunction`.  You can still mock it but it makes the dependencies a lot clearer.

Reply:
Done.